### DISCUSSION TITLE
@field_validator doesn't work for query params

### DISCUSSION THREAD

**Discussion by vunhatchuong** [fastapi/fastapi#12374] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@router.get("")
async def get_all(
    filter: HolidayFilter = Depends(),
    db: AsyncSession = Depends(get_db),
) -> PageOptionalParams[Holiday]:
    return await holiday_crud.filter(
        db,
        None if all(val is None for val in dict(filter).values()) else filter,
    )

class HolidayFilter(SQLModel):
    name: str | None = None
    date: datetime.date | None = None

    @field_validator("date", mode="before")
    @classmethod
    def string_to_date(cls, v: Any) -> Any:
        print("I'm here")
        if isinstance(v, str):
            return None
        return v
```

### Description

When I correctly put the date: `/holiday?date=2024-01-01`, it prints "I'm here" indicating the method runs.

But when I put an invalid date (a string) `/holiday?date=2024-00-00` it doesn't print "I'm here" at all so it doesn't detect that it's a string and returns None.

This is the response:
```json
{
  "detail": [
    {
      "type": "date_from_datetime_parsing",
      "loc": [
        "query",
        "date"
      ],
      "msg": "Input should be a valid date or datetime, month value is outside expected range of 1-12",
      "input": "2024-00-00",
      "ctx": {
        "error": "month value is outside expected range of 1-12"
      }
    }
  ]
}
```

### Operating System

Linux

### FastAPI Version

0.115.0

### Pydantic Version

2.9.2

### Python Version

3.10.15

**Comment by Kfir-G** [comment]
Hey @vunhatchuong, the issue occurs because Pydantic tries to parse the query parameter as a date before your custom `@field_validator` can run, so invalid dates raise an error first. You can fix this by accepting the date as a string and manually validating it in the `@validator` method.

**Reply by vunhatchuong** [reply]
I saw this answer from stackoverflow with Pydantic V1: https://stackoverflow.com/a/76631354

Which still use datetime instead of string for safety. I'm unable to do this on V2.

**Reply by Kfir-G** [reply]
In Pydantic V2, the behavior has changed a bit, but you can still use a similar approach while keeping the safety of datetime. You just need to adjust how you use the validators since `@field_validator` has been replaced by `@model_validator` -> https://github.com/pydantic/pydantic/issues/5245

**Reply by Kfir-G** [reply]
https://stackoverflow.com/questions/76734333/pydantic-v2-field-validator-values-argument-equivalent

**Reply by vunhatchuong** [reply]
Sorry I'm not following, I use the correct `@field_validator()` in my code above like the link but it doesn't work. I can't see the problem

**Comment by Jkrox** [comment, marked as answer]
I'm testing your code with something like this:

```python
import datetime
from typing import Optional, Any

from fastapi import FastAPI, Query
from pydantic import BaseModel, field_validator

class HolidayFilter(BaseModel):
    name: Optional[str] = None
    date: Optional[datetime.date] = None

    @field_validator("date", mode="before")
    @classmethod
    def string_to_date(cls, v: Any) -> Any:
        print("I'm here")
        if isinstance(v, str):
            return None
        return v

app = FastAPI()

@app.get("/")
async def get_all(holiday: HolidayFilter = Query(None)):
    print(holiday)
    return holiday
```

And I got this from terminal:

```
INFO:     Started server process [81970]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     127.0.0.1:41730 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:41730 - "GET /openapi.json HTTP/1.1" 200 OK
I'm here
name='whatever' date=None
INFO:     127.0.0.1:41734 - "GET /?name=whatever&date=2021-00-00 HTTP/1.1" 200 OK
```
Dependencies (uv): 

`pyproject.toml`
```toml
[project]
name = "discussion-12374-fastapi"
version = "0.1.0"
readme = "README.md"
requires-python = ">=3.12"
dependencies = [
    "fastapi>=0.115.0",
    "pydantic>=2.9.2",
    "uvicorn>=0.31.0",
]
```
![image](https://github.com/user-attachments/assets/a8485397-1df8-445c-8f61-8b261c9e8dfe)
---
### Conclusion
I think the issue is over here -> `filter: HolidayFilter = Depends()`.  Hope this helps

**Reply by vunhatchuong** [reply]
I think there's a problem, this makes it so that `date` is always a string.
